### PR TITLE
fix quest content and condition triggers

### PR DESCRIPTION
**Pull request by scooterboo**
## Description
Worked with @Hartie95 to audit and fix quest content and condition triggers.
The main focus was moving away from looking at the passed in params and instead looking at the condition's params.

All changes have been lightly tested. There are none that are completely untested.

## Issues fixed by this PR
Evil Spirits Begone!
image

## Type of changes

- [x] Bug fix
- [x] New feature 
- [x] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

### CODE REVIEW

File: src/main/java/emu/grasscutter/game/quest/conditions/ConditionPackHaveItem.java [hunk 1]
```diff
@@ -19,7 +19,8 @@ public boolean execute(
             int... params) {
         val itemId = condition.getParam()[0];
         val targetAmount = condition.getParam()[1];
-        val checkItem = owner.getInventory().getItemByGuid(itemId);
-        return checkItem != null && checkItem.getCount() >= targetAmount;
+        val checkItem = owner.getInventory().getItemById(itemId);
+        val amount = owner.getInventory().getItemCountById(itemId);
+        return checkItem != null && amount >= targetAmount;
```
Comment (Hartie95) [inline src/main/java/emu/grasscutter/game/quest/conditions/ConditionPackHaveItem.java]: I don't think you still need checkItem if you get the count via inventory

**Review by Hartie95 — APPROVED** [review]

File: src/main/java/emu/grasscutter/game/dungeons/DungeonManager.java [hunk 2]
```diff
@@ -286,23 +286,30 @@ public void startDungeon() {
...
-                .getPlayers()
-                .forEach(player -> player.getPlayerProgress().markDungeonAsComplete(dungeonId));
-
         notifyEndDungeon(true);
         endDungeon(BaseDungeonResult.DungeonEndReason.COMPLETED);
     }
 
+    public void quitDungeon() {
+        notifyEndDungeon(false);
+        endDungeon(BaseDungeonResult.DungeonEndReason.QUIT);
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/dungeons/DungeonManager.java]:
```suggestion
        this.notifyEndDungeon(false);
        this.endDungeon(BaseDungeonResult.DungeonEndReason.QUIT);
```

File: src/main/java/emu/grasscutter/game/dungeons/DungeonManager.java [hunk 3]
```diff
@@ -286,23 +286,30 @@ public void startDungeon() {
...
+    }
+
+    public void failDungeon() {
+        notifyEndDungeon(false);
+        endDungeon(BaseDungeonResult.DungeonEndReason.FAILED);
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/dungeons/DungeonManager.java]:
```suggestion
        this.notifyEndDungeon(false);
        this.endDungeon(BaseDungeonResult.DungeonEndReason.FAILED);
```

File: src/main/java/emu/grasscutter/game/dungeons/DungeonManager.java [hunk 4]
```diff
@@ -286,23 +286,30 @@ public void startDungeon() {
...
                 .getPlayers()
                 .forEach(
                         p -> {
-                            // Trigger the fail event if needed.
-                            if (!successfully) {
+                            // Trigger the fail and success event.
+                            if (successfully) {
+                                var dungeonId = this.getDungeonData().getId();
+                                p.getPlayerProgress().markDungeonAsComplete(dungeonId);
+                            }else{
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/dungeons/DungeonManager.java]:
```suggestion
                            } else {
```

File: src/main/java/emu/grasscutter/game/inventory/Inventory.java [hunk 5]
```diff
@@ -62,6 +65,26 @@ public GameItem getItemByGuid(long id) {
...
+
+    @Nullable
+    public GameItem getItemById(int itemId) {
+        val inventoryTab = getInventoryTabByItemId(itemId);
+        return inventoryTab!=null ? inventoryTab.getItemById(itemId) : null;
+    }
+
+    public int getItemCountById(int itemId){
+        val inventoryTab = getInventoryTabByItemId(itemId);
+        return inventoryTab!=null ? inventoryTab.getItemCountById(itemId) : 0;
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/inventory/Inventory.java]:
```suggestion
        val inventoryTab = this.getInventoryTabByItemId(itemId);
        return inventoryTab != null ? inventoryTab.getItemCountById(itemId) : 0;
```

File: src/main/java/emu/grasscutter/game/inventory/Inventory.java [hunk 6]
```diff
@@ -62,6 +65,26 @@ public GameItem getItemByGuid(long id) {
...
+        if (itemData == null || itemData.getItemType() == null) {
+            return null;
+        }
+        return getInventoryTab(itemData.getItemType());
+    }
+
+    @Nullable
+    public GameItem getItemById(int itemId) {
+        val inventoryTab = getInventoryTabByItemId(itemId);
+        return inventoryTab!=null ? inventoryTab.getItemById(itemId) : null;
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/inventory/Inventory.java]:
```suggestion
        val inventoryTab = this.getInventoryTabByItemId(itemId);
        return inventoryTab != null ? inventoryTab.getItemById(itemId) : null;
```

File: src/main/java/emu/grasscutter/game/quest/QuestManager.java [hunk 7]
```diff
@@ -279,7 +279,7 @@ public GameMainQuest getMainQuestById(int mainQuestId) {
     }
 
     public GameMainQuest getMainQuestByTalkId(int talkId) {
-        int mainQuestId = GameData.getQuestTalkMap().getOrDefault(talkId, talkId / 100);
+        int mainQuestId = GameData.getQuestTalkMap().getOrDefault(talkId, 0);
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/quest/QuestManager.java]:
```suggestion
        var mainQuestId = GameData.getQuestTalkMap().getOrDefault(talkId, 0);
```

File: src/main/java/emu/grasscutter/game/quest/content/ContentCompleteAnyTalk.java [hunk 8]
```diff
@@ -2,22 +2,32 @@
...
             GameQuest quest, QuestData.QuestContentCondition condition, String paramStr, int... params) {
-        return Stream.of(condition.getParamStr().split(","))
+        var conditionTalk = Arrays.stream(condition.getParamStr().split(","))
                 .mapToInt(Integer::parseInt)
-                .anyMatch(
-                        talkId ->
-                                GameData.getTalkConfigDataMap().get(params[0]) != null && talkId == params[0]);
+                .toArray();
+
+        for(var talkId : conditionTalk) {
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/quest/content/ContentCompleteAnyTalk.java]:
```suggestion
        for (var talkId : conditionTalk) {
```

File: src/main/java/emu/grasscutter/game/quest/content/ContentCompleteAnyTalk.java [hunk 9]
```diff
@@ -2,22 +2,32 @@
...
+            val checkMainQuest = quest.getOwner().getQuestManager().getMainQuestByTalkId(talkId);
+            if (checkMainQuest == null) {
+                if (talkId == params[0]) return true;
+                continue;
+            }
+            val talkData = checkMainQuest.getTalks().get(talkId);
+            if(talkData != null) {
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/quest/content/ContentCompleteAnyTalk.java]:
```suggestion
            if (talkData != null) {
```

File: src/main/java/emu/grasscutter/game/quest/content/ContentCompleteAnyTalk.java [hunk 10]
```diff
@@ -2,22 +2,32 @@
...
-                                GameData.getTalkConfigDataMap().get(params[0]) != null && talkId == params[0]);
+                .toArray();
+
+        for(var talkId : conditionTalk) {
+            val checkMainQuest = quest.getOwner().getQuestManager().getMainQuestByTalkId(talkId);
+            if (checkMainQuest == null) {
+                if (talkId == params[0]) return true;
+                continue;
+            }
+            val talkData = checkMainQuest.getTalks().get(talkId);
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/quest/content/ContentCompleteAnyTalk.java]:
```suggestion
            }

            val talkData = checkMainQuest.getTalks().get(talkId);
```

**Review by KingRainbow44 — APPROVED** [review]

**Pull request merged**